src/OrleansDockerUtils/DockerMembershipOracle.cs

```diff
@@ -0,0 +1,468 @@
+                }
+
+                // Identify removed silos.
+                foreach (var silo in silos)
+                {
+                    // Never remove self.
+                    if (silo.Key.Equals(SiloAddress)) continue;
+
+                    // Do not remove silos which were present in the update.
+                    if (silo.Value.Refreshed)
+                    {
+                        // Reset the refresh flag.
+                        silo.Value.Refreshed = false;
+                        continue;
+                    }
+
+                    // The silo was not present in the update, remove it.
+                    if (removed == null) removed = new HashSet<SiloAddress>();
+                    removed.Add(silo.Key);
+                }
+
+                // If any silos were removed, 
+                if (removed != null)
+                {
+                    foreach (var silo in removed)
+                    {
+                        silos.Remove(silo);
+                    }
+                }
+            }
+
+            // If anything was updated, clear the cache before notifying clients.
+            if (added != null || removed != null)
+            {
+                // Clear the caches.
+                ClearCaches();
+
+                // If any silos were added or removed, notify subscribers of the new status.
+                var siloStatusListeners = subscribers.Values.ToList();
+                if (added != null)
+                {
+                    foreach (var address in added)
+                    {
+                        NotifySubscribers(address, SiloStatus.Active, siloStatusListeners);
+                    }
+                }
+
+                if (removed != null)
+                {
+                    foreach (var address in removed)
+                    {
+                        NotifySubscribers(address, SiloStatus.Dead, siloStatusListeners);
+                    }
+                }
+            }
+        }
+
+        private async Task RefreshAsync()
+        {
+            try
+            {
+                await resolver.Refresh();
+
+                lastRefreshTime = DateTime.UtcNow;
+            }
+            catch (Exception exception)
+            {
+                log?.Warn(
+                    (int)ErrorCode.Docker_MembershipOracle_ExceptionRefreshingSilos,
+                    "Exception refreshing silos.",
+                    exception);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Clears the cached data.
+        /// </summary>
+        private void ClearCaches()
+        {
+            activeSilosCache = null;
+            allSilosCache = null;
+            multiClusterSilosCache = null;
+        }
+
+        /// <summary>
+        /// Updates the status of this silo.
+        /// </summary>
+        /// <param name="status">The updated status.</param>
+        private void UpdateStatus(SiloStatus status)
```

Comment:
It does not propagate the new status of this silo to other silos (yes, it does update local subscribers).

When a silo goes lets say to terminating state (just an example), we need to notify all other silos about it. Its not enough to notify all local components. If we can't do distributed silo states, that's a major gap that we are opening compared to the main MBR Oracle and we need to think through the implications.